cpp/src/interop/arrow_to_cudf.cu

```diff
@@ -0,0 +1,302 @@
+    case arrow::Type::type::INT64: return data_type(type_id::INT64);
+    case arrow::Type::type::UINT8: return data_type(type_id::UINT8);
+    case arrow::Type::type::UINT16: return data_type(type_id::UINT16);
+    case arrow::Type::type::UINT32: return data_type(type_id::UINT32);
+    case arrow::Type::type::UINT64: return data_type(type_id::UINT64);
+    case arrow::Type::type::FLOAT: return data_type(type_id::FLOAT32);
+    case arrow::Type::type::DOUBLE: return data_type(type_id::FLOAT64);
+    case arrow::Type::type::DATE32: return data_type(type_id::TIMESTAMP_DAYS);
+    case arrow::Type::type::DATE64: return data_type(type_id::TIMESTAMP_MILLISECONDS);
+    case arrow::Type::type::STRING: return data_type(type_id::STRING);
+    case arrow::Type::type::DICTIONARY: return data_type(type_id::DICTIONARY32);
+    case arrow::Type::type::LIST: return data_type(type_id::LIST);
+    default: CUDF_FAIL("Unsupported type_id conversion to cudf");
+  }
+}
+
+namespace {
+/**
+ * @brief Functor to return pair of column and column view for a corresponding arrow array. column
+ * is formed from buffer underneath the arrow array, and if the array is resized or has an offset,
+ * that is taken care by the column view provided along with it
+ */
+struct dispatch_to_cudf_column {
+  template <typename T>
+  std::pair<std::unique_ptr<column>, column_view> operator()(arrow::Array const& array,
+                                                             data_type type,
+                                                             bool skip_mask,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             cudaStream_t stream)
+  {
+    auto data_buffer = array.data()->buffers[1];
+    auto num_rows    = data_buffer->size() / sizeof(T);
+    auto has_nulls   = skip_mask ? false : array.null_bitmap_data() != nullptr;
+    auto col         = make_fixed_width_column(
+      type, num_rows, has_nulls ? mask_state::UNINITIALIZED : mask_state::UNALLOCATED, stream, mr);
+    auto mutable_column_view = col->mutable_view();
+    CUDA_TRY(cudaMemcpyAsync(mutable_column_view.data<void*>(),
+                             data_buffer->data(),
+                             data_buffer->size(),
+                             cudaMemcpyHostToDevice,
+                             stream));
+    if (has_nulls) {
+      CUDA_TRY(cudaMemcpyAsync(mutable_column_view.null_mask(),
+                               array.null_bitmap_data(),
+                               array.null_bitmap()->size(),
+                               cudaMemcpyHostToDevice,
+                               stream));
+    }
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(array.offset()),
+                          static_cast<size_type>(array.offset() + array.length())));
```

Comment:
I see, this function returns a `column` and `column_view` because it copies the entire `arrow::Array` data buffer even if it has an offset. This seems wasteful.

Why not only copy the data in the `arrow::Array` from `[offset, size)`?

Reply:
In case of string or list, the offsetting the child column doesn't make any sense. And for which you have to completely form the column and then take the offset.

Reply:
Else I can  handle fixed width completely different and child columns of compound times completely different.

Reply:
I'm not following what the problem is.

Reply:
I had started with some approach with offsets in mind, but with end changes I think it may not be required. Will just copy from offset to size as you mentioned.

Reply:
I'm not sure if this is handled, but in case not -- we need to be sure to rebase the codes from the offset position when copying sliced String, and List columns to/from Arrow:
```
// original Arrow::StringArray or cudf::column
string_col = { offset: 0, offsets: [0,1,2,3], data: ["abcd"] }

// sliced Arrow::StringArray or cudf::column (offsets sliced, data untouched)
offset_col = { offset: 2, offsets: [2,3], data: ["abcd"] }

// result of to_arrow or from_arrow with `offset_col` input
result_col = { offset: 0, offsets: [0,1], data: ["cd"] }
```

Reply:
I am copying whole offsets and data and then slicing it in cudf, https://github.com/rapidsai/cudf/blob/0794298c5dbd2a01fa0130aa10a7565f3dbffaea/cpp/src/interop/from_arrow.cu#L173